rsocket-routing-broker-spring/src/main/java/io/rsocket/routing/broker/spring/BrokerAutoConfiguration.java

```diff
@@ -110,6 +114,20 @@ else if (log.isWarnEnabled()) {
 		}
 	}
 
+
+	@Bean
+	@ConditionalOnProperty(prefix = BROKER_PREFIX, name = "default-load-balancer", havingValue = "weighted")
+	public RSocketServerCustomizer customizer() {
+		// TODO: should always be installed if algorithm is selected based on the given
+		//       tags
+		return rSocketServer -> rSocketServer.interceptors(ir -> ir.forRequester((Function<RSocket, RequestInterceptor>) rSocket -> {
+			final WeightedStatsRequestInterceptor weightedStatsRequestInterceptor =
+					new WeightedStatsRequestInterceptor();
+			ir.forRequester((RSocketInterceptor) rSocket1 -> new WeightedStatsAwareRSocket(rSocket1, weightedStatsRequestInterceptor));
+			return weightedStatsRequestInterceptor;
```

Comment:
On closer look this doesn't look entirely alright to me. An `RequestInterceptor`, scoped to a specific `RSocket`, that's installing a more global `RSocketInterceptor` and wouldn't that interceptor then apply to every subsequent `RSocket`? That means after a few `RSocket` connections, an `RSocket` gets as many `WeightedStatsAwareRSocket` decorators which would probably still work because the latest should be on top, or maybe not if connections are concurrently established..

I think we need to reverse this somehow such that there is only one `RSocketInterceptor` which is installed on startup and it contains a `Map<RSocket, WeightedStatsRequestInterceptor>` so when it is called it knows how to decorate the `RSocket`. Similar to how it works by default in `WeightedStatsLoadbalanceStrategy` but here adding a proxy instead.

Reply:
> On closer look, this doesn't look entirely alright to me. A `RequestInterceptor`, scoped to a specific RSocket, that's installing a more global RSocketInterceptor, and wouldn't that interceptor then apply to every subsequent RSocket?

The interceptor function will be called whenever the new connection is accepted, which means that every separate requester will have its own interceptor.

> That means after a few RSocket connections, an RSocket gets as many WeightedStatsAwareRSocket decorators, which would probably still work because the latest should be on top, or maybe not if connections are concurrently established.

The RequestInterceptor function is applied only once for the source rsocket and does not apply to every other decorator. Also, the function is called when the RSocketRequester is constructed, so the interceptor added at line 126 will be the very last, so we will be sure that the RSocket is given SocketAcceptor is the `WeightedStatsAwareRSocket` one.

> I think we need to reverse this somehow such that there is only one RSocketInterceptor installed on startup. It contains a Map<RSocket, WeightedStatsRequestInterceptor>, so when it is called, it knows how to decorate the RSocket. Similar to how it works by default in WeightedStatsLoadbalanceStrategy, but here adding a proxy instead.

We can always have a registry. It is easy to create such ( as we have done for the client-side ). But, from my standpoint, having a single access object is not that scalable. At least 1 or 2 volatile reads may decrease performance significantly, especially when a broker should efficiently rout requests. On the other hand, the broker is not a client-facing library. Hence we are free to do any hacks to achieve better performance. Thus, we can keep such lambda in a lambda to avoid extra `ConcurrentMap#get` operation and then  freely cast rsocket to WeightedStats since we know in advance that this is the top wrapper